### PR TITLE
Start of a script to explore which solutions an overparameterized model finds

### PR DESCRIPTION
**Pull request by ChickenProp**
I'm not sure I have results yet, but I'm training an overparameterized model and then measuring its broadness. The model is using nine neurons to solve xor. By default just one hidden layer, but there's also one with a 5-neuron layer and a 4-neuron layer in the code.

I'm not sure how to interpret the output. It's a list of deltas, so a running sum of the list would give the size of the basin at different distances? But I'm not sure how to interpret that value either.

A thing someone suggested to me was to graph the output as a heatmap, since it's 2d input and 1d output. Doing that for several seeds feels to me like it would get at the question we're trying to answer? I don't think it would be very difficult.

I'm also just generally not confident that the script correctly does what I think it does, in particular the subclassing of `BroadnessMeasurer`. Here's the output I currently get at different seeds:

```
0:
[ 2.01118723e-04  1.99670044e-04  4.14478190e-04  2.58260722e-05
  2.08504142e-04 -8.12837078e-05  6.79794723e-04  4.89184829e-04
 -2.56717853e-04]

1:
[ 2.95658515e-04  2.52063453e-04  1.93877040e-04  7.12543897e-05
  6.07903406e-04 -2.53398756e-04  3.90943979e-04  8.77720608e-05
  8.48936580e-05]

2:
[ 0.00015619  0.00027915  0.00014813  0.00033873  0.00031212 -0.00050297
  0.00108656 -0.00016611 -0.00032301]

3:
[ 1.40713393e-04  2.22427241e-04  5.42331432e-04  3.47921393e-05
  3.69817737e-04  3.88683834e-05  2.73165298e-04 -1.56782846e-04
  9.02565662e-04]

4:
[ 2.54370926e-04  4.17749549e-05  4.55194886e-04  3.89253605e-04
  5.35306061e-04 -4.06768425e-04  4.22502403e-04  2.22372937e-04
  2.59280385e-04]
```

and here's what I get if I switch to `BroadnessMeasurer` instead of `OurBroadnessMeasurer`:

```
0:
[ 2.14634836e-04  3.83138657e-04  8.91610980e-05 -2.21960247e-04
  1.21752918e-04 -2.44839489e-04  5.14343381e-04  1.37908906e-03
 -1.68861449e-04]

1:
[ 6.47072494e-04 -2.35565603e-03  1.79990083e-03  1.60422921e-03
  9.96813178e-05 -7.37029314e-04  4.13085520e-04 -4.14356589e-04
  5.66322207e-03]

2:
[ 0.0007683  -0.00088865 -0.00074181 -0.00056642  0.00499493 -0.00447496
 -0.00014621  0.00275478 -0.00043517]

3:
[-9.42274928e-04 -4.82797623e-06 -1.52532756e-04  1.97724253e-03
  1.26041770e-03  2.66335011e-03 -4.72096205e-03  8.28191638e-05
  1.76514685e-03]

4:
[ 0.00099557 -0.00089882 -0.00078551  0.00128097  0.00110032  0.00047836
 -0.0022034   0.00345393 -0.00072984]
```

### CODE REVIEW

**Comment by LuanAdemi** [comment]
I'm gonna check your script for bugs. The idea with the heatmaps sounds good. And you could check out the discord chat between @Davanchama and Avery for getting an idea on how to interpret the output (# ave-luan-david).

File: experiments/overparameterized_broadness.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+class Net_54(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.fc1 = nn.Linear(2, 5)
+        self.fc2 = nn.Linear(5, 4)
+        self.fc3 = nn.Linear(4, 1)
+
+    def forward(self, x):
+        # F.relu doesn't work here (best loss is 0.17).
```
Comment (LuanAdemi) [inline experiments/overparameterized_broadness.py]: Just FYI: The reason relu performs so badly here is due to the characteristic of it. You can think of the difference between sigmoid and relu in this context in terms of degrees of freedom for predicting 0 and 1 (which is our goal here since we are training our model to perform a binary operation). Since Sigmoid maps big numbers to 1 and small numbers to zero, in contrast to relu, which only maps small numbers to zero, the network is able to use the mapping of the sigmoid to collapse both big and small numbers into equivalence classes which enable a bigger space of possible weight configurations, hence more solutions.

In general the following holds true for every architecture:
ReLU is always used for regression tasks, Sigmoid for binary classification tasks

**Review by LuanAdemi — APPROVED** [review]
The script seems good to me.

**Comment by LuanAdemi** [comment]
I am technically able to merge this, but would you be able to move the script into experiments/research/overparameterized-model/ ?

File: experiments/overparameterized_broadness.py (hunk 1 above)
Comment (ChickenProp) [reply experiments/overparameterized_broadness.py]: Thanks. That... kinda makes sense, but then I'm surprised [silu](https://pytorch.org/docs/stable/generated/torch.nn.SiLU.html) works better (or at least faster) than [sigmoid](https://pytorch.org/docs/stable/generated/torch.nn.Sigmoid.html). Silu also maps small numbers to 0 and big numbers to themselves, I would guess the main advantage it has over relu here is that it's at least capable of generating negative numbers?

I played around with https://dcato98.github.io/playground/ a bit to see if I could get a feel for what's going on, but it always uses tanh for the final output on classification problems and doesn't have anything like the xor dataset for regression. Plus the input is in [-1, 1] instead of {0, 1} which seems like it could make a big difference here. And it doesn't have silu.

**Comment by ChickenProp** [comment]
Yep, done :)

I'm curious if you have ideas about why the standard `BroadnessMeasurer` gets different results from this script's `OurBroadnessMeasurer`?

**Comment by LuanAdemi** [comment]
Hmm. Maybe @Davanchama can help. He wrote the BroadnessMeasurer class.
(He is currently looking into it)

I'm gonna merge if we resolve your open questions, if that's ok with you.

File: experiments/overparameterized_broadness.py (hunk 1 above)
Comment (LuanAdemi) [reply experiments/overparameterized_broadness.py]: You are probably correct here. It all depends on the landscape of the loss functions image space. Different non- linearitys enable different solutions (like e.g. using negative activation). Looking at the latest research output for superpositions, this might also has a big impact on your result, since your models might inhibit different amounts of superpositions (like in a model of the same size might me able to encode more features depending on the activation function). Might be something we could look into after you are done with your current setup. For now I would suggest sticking with one activation function throughout your experiment.

**Comment by Davanchama** [comment]
Hey! I think I found the bug in my last commit. We developed this measurer for many different model architectures and while adapting the class to these different architectures and IO dimensions, one little .view(-1) survived at the wrong place. Removing it resolved the differences between batching and not batching for me. I hope it also does for you :)
(And by the way, thanks for pointing this out, this was a huge bug. In a few days I will test this class on different datasets to make sure this .view() was not a quick fix for a specific dataset.)

**Comment by LuanAdemi** [comment]
@ChickenProp ready to merge?

**Comment by ChickenProp** [comment]
Thanks @Davanchama! Unfortunately, I still get different results with `OurBroadnessMeasurer` versus `BroadnessMeasurer`. The former still gets the same results as it did before, but the latter has changed:

```
0:
[ 0.00024168  0.00023018 -0.00010062  0.00053654  0.00018119 -0.00028401
  0.00186747 -0.00133161  0.00150165]

1:
[ 0.00032748  0.00013177  0.00028681 -0.00012146  0.00051109  0.00050925
 -0.00018844  0.00074433 -0.00042814]

2:
[2.04037390e-04 1.97912108e-04 1.90232368e-04 2.53006521e-04
 9.12579475e-05 1.59446625e-04 6.02603895e-04 9.32585325e-04
 4.36844370e-05]

3:
[ 0.00011847  0.00030322  0.0003507   0.00047351 -0.00053798  0.00066225
  0.00052862 -0.00061835  0.00191112]

4:
[ 2.77112638e-04  3.32914489e-04  3.42315338e-04 -1.58084978e-05
  7.09976382e-04 -3.89396891e-04  4.48546857e-04  6.59014602e-04
  5.66759563e-04]
```

I haven't looked closely into where the differences come from yet, but I've pushed the merged commit I was comparing with. I did have to change the device in `BroadnessMeasurer` back to `cpu` to compare them, since I don't have an nvidia GPU.

**Comment by ChickenProp** [comment]
@LuanAdemi happy to merge when we figure this out.

**Comment by ChickenProp** [comment]
Okay, I finally made time to look at this again. It seems (at least part of?) the difference is caused by getting a different result when I run the model on one data point versus multiple data points at a time.

For example, on the trained model, `print(net(data))` gives

```
tensor([[8.4639e-06],
        [9.9999e-01],
        [9.9999e-01],
        [1.2875e-05]], grad_fn=<AddmmBackward0>)
```

but `for i in range(4): print(net(data[i]))` gives

```
tensor([8.4043e-06], grad_fn=<AddBackward0>)
tensor([1.0000], grad_fn=<AddBackward0>)
tensor([1.0000], grad_fn=<AddBackward0>)
tensor([1.2815e-05], grad_fn=<AddBackward0>)
```

where all of those lines are slightly different. If I make it `data[i].view(1, 2)` the results are the same except wrapped in an extra `[ ]`, so it seems to be maybe something to do with the number of inputs, not the dimension.

My naive guess would be this is just numeric instability in matrix multiplication algorithms or something, and at this scale either result can be considered basically correct? And if that's right, then the difference between the results with `BroadnessMeasurer` and with `OurBroadnessMeasurer` is basically noise, and if it's too large then we should bump up num_iters. (And this would also apply if the difference between runs with the same measurer but different seeds is too large. But I don't know what "too large" would mean in this context, I don't know what sort of scale of numbers to expect here.)

Does that sound right?

**Comment by Davanchama** [comment]
Hm this is interesting. So if this is the reason for the different results, I would be satisfied with leaving this as it is, as (as you said) a high number of num_iters should make these inaccuracies meaningless. Did you also get different losses when losses get high or do the errors stay?

**Comment by ChickenProp** [comment]
(I assume you mean "when num_iters gets high"?)

If I move it up to 10000, which gives about a minute of runtime for me, seeds 0-4 give the following results (I've multiplied them all by 1000):

```
OurBroadnessMeasurer:
[0.23108 0.24295 0.23768 0.22039 0.22694 0.26352 0.22125 0.26906 0.23632]
[0.24047 0.23806 0.25814 0.23716 0.21712 0.27636 0.21262 0.25544 0.22893]
[0.22745 0.22398 0.20939 0.23007 0.21185 0.2442  0.21189 0.21314 0.22917]
[0.25547 0.26182 0.27432 0.24932 0.24908 0.30844 0.23023 0.29392 0.26077]
[0.27862 0.26855 0.29785 0.26054 0.28727 0.30004 0.29576 0.26454 0.24938]

BroadnessMeasurer:
[0.2357  0.23159 0.24921 0.22994 0.23662 0.23618 0.22095 0.25829 0.2277 ]
[0.24147 0.23956 0.23881 0.24323 0.23524 0.25706 0.20385 0.27091 0.25953]
[0.22688 0.21922 0.22731 0.22975 0.1952  0.25209 0.22365 0.20842 0.20409]
[0.25879 0.26889 0.25995 0.25106 0.29641 0.22841 0.27292 0.2382  0.26923]
[0.27848 0.2821  0.2813  0.26275 0.26579 0.30881 0.26761 0.28333 0.25016]
```

Eyeballing, I'd say those look similar enough that I guess there's nothing left to explain?

In light of this I've removed `OurBroadnessMeasurer` from the script and left num_iters high.

**Comment by LuanAdemi** [comment]
So increasing `num_iters` and therefore the sample size per standard deviation resolved the value difference? Seems like this is finally getting merged. 😀 🥳
